### PR TITLE
Term search

### PR DESCRIPTION
**Pull request by airenzp**
## Description

This PR adds the TermTypeSearch component. Now everytime the tree is loaded the tabs may be shown if needed. Once a tab is selected a corresponding view is shown to select a term. The search input was customized to only search for items of a type. On each use case the component should allow the expected tabs. I worked in this together with @siosonel. Please check.

## Checklist

[Check each task](https://github.com/stjude/proteinpaint/wiki/Pull-Request-Checklist) that has been performed or verified to be not applicable.
- [x] Tests: added and/or passed unit and integration tests, or N/A
- [x] Todos: commented or documented, or N/A
- [x] Notable Changes: updated release.txt, prefixed a commit message with "fix:" or "feat:", added to an internal tracking document, or N/A

### CODE REVIEW

**Comment by gavrielm** [comment]
I know this is still a draft PR, but I noticed that `SNP List` and `SNP Locus` both error out in the [sjlife portal dictionary](http://localhost:3000/?noheader=1&mass={%22genome%22:%22hg38%22,%22dslabel%22:%22SJLife%22,%22nav%22:{%22activeTab%22:1},%22plots%22:[{%22chartType%22:%22dictionary%22}]}) when I try to submit an input.

Also, the UIs for SNP List and SNP Locus are written for regression analysis. They will likely need to be modified for rendering in data dictionary.

**Comment by xzhou82** [comment]
can keep snplst and snplocus as-is and just make them to work for regression and download

**Comment by karishma-gangwani** [comment]
Will the omni search box have specific default terms based on the data type in the future? Now it says 'search term' for both.

**Comment by gavrielm** [comment]
> I just pushed a fix for the dictionary use case, please check again.

Great thanks. In regression, this intermediate menu should be removed:

image

**Review by karishma-gangwani — APPROVED** [review]
It does look nice with the terms segregated according to data types. Should we just label 'Data Variables' as 'Variables' since the button does say `Data dictionary' ? Just a thought.

**Review by gavrielm — DISMISSED** [review]
Tested in sjlife, mbmeta, and pnet and all work well

**Comment by creilly8** [comment]
This is minor. If I search for something in one tab, do not hit enter, then switch to another tab, the search results remain: 
Screenshot 2024-04-11 at 2 19 12 PM

Also minor: I noticed the input must be at least 3 characters before a search result appears. That works well for the dictionary terms because they're long, but what about the mutations, genes, etc? They're really short.

**Comment by creilly8** [comment]
@airenzp After looking at your PR, it occurs to me we never include a `No Result` message for non-matches for any of our inputs. We probably do that going forward. What do you think?

**Comment by airenzp** [comment]
> It does look nice with the terms segregated according to data types. Should we just label 'Data Variables' as 'Variables' since the button does say `Data dictionary' ? Just a thought.

@karishma-gangwani  this component is shown everywhere there is a tree with more than one choice, not just under the mass UI dictionary.

**Comment by airenzp** [comment]
> I have two questions.
> 
> 1. Will the omni search box have specific default terms based on the data type in the future? Now it says 'search term' for both.
> 2. Would this be also supported for overlay and divide by later?

Thanks for the review.  Now I show a different placeholder when changing tabs. And it is customized now to show geneVariant on the overlay and divide by, if you find something that before was possible missing let me know!

**Comment by airenzp** [comment]
> This is minor. If I search for something in one tab, do not hit enter, then switch to another tab, the search results remain: Screenshot 2024-04-11 at 2 19 12 PM
> 
> Also minor: I noticed the input must be at least 3 characters before a search result appears. That works well for the dictionary terms because they're long, but what about the mutations, genes, etc? They're really short.

Thanks @creilly8 . now when changing tabs the search is cleared. Thanks for pointing this out. I also noticed we need to type 3 characters to find a gene. @xzhou82 should we ask for less? like 2 chars instead?

**Comment by karishma-gangwani** [comment]
> > I have two questions.
> > 
> > 1. Will the omni search box have specific default terms based on the data type in the future? Now it says 'search term' for both.
> > 2. Would this be also supported for overlay and divide by later?
> 
> Thanks for the review. Now I show a different placeholder when changing tabs. And it is customized now to show geneVariant on the overlay and divide by, if you find something that before was possible missing let me know!

works great! Thanks Airen

**Comment by xzhou82** [comment]
there's no `omnisearch`. dictionary search will only turn up dict terms

**Comment by xzhou82** [comment]
gene search should allow single letter (gene `T`). Airen can search.js accomondate that?

**Comment by xzhou82** [comment]
for searching genevariant/geneexp/etc, best to just use `dom/genesearch.js`

File: server/src/searchSNP.js [hunk 1]
```diff
@@ -14,6 +14,9 @@
 .lst[ str ]
 	used for byName
 */
+
+import { query_bigbed_by_coord } from './utils.js'
+
 export async function searchSNP(q, genome) {
```
Comment (xzhou82) [inline server/src/searchSNP.js:20]: looks like this func is now exported by routes/snp.ts. may dedup

Comment (airenzp) [reply server/src/searchSNP.js]: Done

**Comment by xzhou82** [comment]
snp search works after dedup. thanks. please fix minor issues:
- at [violin](http://localhost:3000/?noheader=1&mass={%22dslabel%22:%22SJLife%22,%22genome%22:%22hg38%22,%22plots%22:[{%22chartType%22:%22summary%22,%22term%22:{%22id%22:%22hrtavg%22,%22q%22:{%22mode%22:%22continuous%22}}}],%22nav%22:{%22activeTab%22:1}}), open Overlay breaks
- snplst and snplocus tabs are still shown, e.g. [sjlife cuminc overlay](http://localhost:3000/?noheader=1&mass={%22dslabel%22:%22SJLife%22,%22genome%22:%22hg38%22,%22plots%22:[{%22chartType%22:%22cuminc%22,%22settings%22:{%22cuminc%22:{}},%22term%22:{%22id%22:%22Cardiovascular%20System%22,%22q%22:{%22breaks%22:[1]}}}],%22nav%22:{%22activeTab%22:1}}), and sjlife download (select Dictionary variables first then see the tabs). please hardcode in termTypeSearch to hide them for now
-

**Comment by airenzp** [comment]
> snp search works after dedup. thanks. please fix minor breaking issues:
> 
> * at [violin](http://localhost:3000/?noheader=1&mass=%7B%22dslabel%22:%22SJLife%22,%22genome%22:%22hg38%22,%22plots%22:%5B%7B%22chartType%22:%22summary%22,%22term%22:%7B%22id%22:%22hrtavg%22,%22q%22:%7B%22mode%22:%22continuous%22%7D%7D%7D%5D,%22nav%22:%7B%22activeTab%22:1%7D%7D), open Overlay breaks
> 
> i will create a new issue with non-breaking things for you to fix in next pr. thanks

Thanks. I added the violin use case now.

**Review by xzhou82 — APPROVED** [review]

**Pull request merged**